arangod/Aql/SortBlock.cpp

```diff
@@ -21,256 +21,616 @@
-                  // If the following does not work, we will create a
-                  // few unnecessary copies, but this does not matter:
-                  cache.emplace(a, a);
-                }
-              }
-            } else {
-              // simple value, which does not need ownership transfer
-              TRI_IF_FAILURE("SortBlock::doSortingCache") {
-                THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-              }
-              TRI_IF_FAILURE("SortBlock::doSortingNext1") {
-                THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-              }
-              TRI_IF_FAILURE("SortBlock::doSortingNext2") {
-                THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-              }
-              next->setValue(i, j, a);
-              _buffer[coords[count].first]->eraseValue(coords[count].second, j);
-            }
-          }
-        }
+        ::stealRow(cache, nrRegs, _heapBuffer.front(), _rows[count], next, i);
         count++;
       }
       cache.clear();
     }
-  } catch (...) {
-    for (auto& x : newbuffer) {
+
+    // pre-emptively cleanup memory from heap buffer (also done in destructor)
+    releaseHeapBuffer();
+
+    return TRI_ERROR_NO_ERROR;
+  }
+
+  bool empty() const override { return _buffer.empty() && _heapBuffer.empty(); }
+
+ private:
+  arangodb::Result pushRow(size_t sRow) {
+    using arangodb::aql::AqlItemBlock;
+    using arangodb::aql::AqlValue;
+    using arangodb::aql::RegisterId;
+
+    if (_rowsPushed >= _limit && _cmpInput(0, sRow)) {
+      // skip row, already too low in sort order to make it past limit
+      return TRI_ERROR_NO_ERROR;
+    }
+
+    AqlItemBlock* srcBlock = _buffer.front();
+    TRI_ASSERT(srcBlock != nullptr);
+    AqlItemBlock* dstBlock = _heapBuffer.front();
+    TRI_ASSERT(dstBlock != nullptr);
+    size_t dRow = _rowsPushed;
+
+    if (_rowsPushed >= _limit) {
+      // pop an entry first
+      std::pop_heap(_rows.begin(), _rows.end(), _cmpHeap);
+      dRow = _rows.back();
+      eraseRow(dRow);
+      _rows.pop_back();
+    }
+
+    TRI_IF_FAILURE("SortBlock::doSortingInner") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
+
+    RegisterId nrRegs = srcBlock->getNrRegs();
+    ::stealRowNoCache(nrRegs, srcBlock, sRow, dstBlock, dRow);
+    _rows.emplace_back(dRow);
+    ++_rowsPushed;
+
+    // now insert copy into heap
+    std::push_heap(_rows.begin(), _rows.end(), _cmpHeap);
+
+    return TRI_ERROR_NO_ERROR;
+  }
+
+  void eraseRow(size_t row) {
+    arangodb::aql::AqlItemBlock* block = _heapBuffer.front();
+    arangodb::aql::RegisterId const nrRegs = block->getNrRegs();
+    for (size_t i = 0; i < nrRegs; i++) {
+      block->destroyValue(row, i);
+    }
+  }
+
+  void ensureHeapBuffer(arangodb::aql::AqlItemBlock* src) {
+    if (_heapBuffer.empty()) {
+      arangodb::aql::RegisterId const nrRegs = src->getNrRegs();
+      arangodb::aql::AqlItemBlock* newBlock = _allocate(_limit, nrRegs);
```

Comment:
This and the following looks a bit overengineered...
Why not capture `newBlock` in an `std::unique_ptr` instead, remove all the try...catch stuff and release `newBlock` in case we want to transfer ownership?